fuel-database/src/lib.rs

```diff
@@ -0,0 +1,368 @@
+use crate::{
+    in_memory::transaction::MemoryTransactionView,
+    transactional::DatabaseTransaction,
+};
+use fuel_core_interfaces::common::error::InterpreterError;
+use serde::{
+    de::DeserializeOwned,
+    Serialize,
+};
+use std::{
+    fmt::{
+        self,
+        Debug,
+        Formatter,
+    },
+    marker::{
+        PhantomData,
+        Send,
+    },
+    sync::Arc,
+};
+use thiserror::Error;
+
+pub use error::*;
+
+mod error;
+pub mod in_memory;
+#[cfg(feature = "rocksdb")]
+pub mod rocks_db;
+pub mod tables;
+pub mod transactional;
+
+/// Database tables column ids.
+#[repr(u32)]
+#[derive(
+    Copy, Clone, Debug, strum_macros::EnumCount, PartialEq, Eq, enum_iterator::Sequence,
+)]
+pub enum Column {
```

Comment:
Moving this stuff will affect https://github.com/FuelLabs/fuel-core/pull/796 =( But I will help with it

Reply:
This is an example of why I've been hesitant towards creating a mega-db crate. This shared set of columns seemingly breaks our attempt of having a clean separation of concerns. While implementing mappables and service db ports from the crate perspective is nice, we still end up with a cross-service dependency through these shared columns. Any service that needs to change columns will cascade into potentially breaking adapter changes in all the other crates (even if they had no real logical changes).

While this may not seem like a big issue now when compared to the benefits of sharing the database when considering the executor migration or cross-service, we now open the possibility for implementers of new crates to just start throwing what they need into the shared database crate instead of using their own local ports without really being aware of any difference.

By allowing fuel-core to be responsible for aggregating the database ports from services and having full control over how they are implemented, it resolves the following situations:
 - Allowing custom impls / fine tuning of database traits in the context of a fuel-core node. E.g. using separate instances of rocksdb for merkle tree updates vs plain contract state for performance.
 - Prevent devs from arbitrarily updating a shared database crate whenever they need something custom for their use-case, causing unintended side-affects for other projects using our service crates in different contexts (e.g. putting our block executor into a parachain or cosmos app).

I get how some set of shared data interfaces would be beneficial for things like the executor migration or cross-service txs. An alternative approach to solving these problems without fuel-database or fuel-core-interfaces could be:

- Create an alternative to `DatabaseTransaction` that is entirely trait based. The new trait should have similar AsRef/Deref functionality as before allowing for casts to the db port of each respective service using it.
  - This could be in a new crate as long as it's only highly generic database traits that don't know anything about the underlying services or what their requirements/mappables/columns are. 
- Then services like the executor would just use its own db port trait which defines the mappings it needs, and also use the new trait for database transactions.

This way fuel-core has ultimate control over how db calls get implemented and services don't share any dependency on each other's database requirements.

Reply:
I’ve read through this a few times and it’s not really landing. Can you provide a small example of what a trait based transaction could like like?

Reply:
Is the idea to have some kind of custom functions for each request on a service? So instead of using columns and keys we have a getter function for every piece of data we need?

Reply:
## Purpose of this PR

I want to highlight one more time that this change only moves stuff to make the review as simple as possible. Real refactoring to be a crate of database primitives will be done in follow-ups. There are two variants of how to create this crate:
- Extract the whole logic into one crate and break it down(to downstream crates) with new changes after.
- Define new primitives in this new crate. After, we will start refactoring existing logic to use those primitives.

I selected the first approach because it allows refactoring with the right primitives that will be easily renamed/replaced later. And it will parallelize refactoring(unblock in some cases because anyone can contribute to the `fuel-database` crate) to use ports/adaptors and new database primitives. 

> This could be in a new crate as long as it's only composed of highly generic database traits that don't know anything about the underlying services or what their requirements/mappables/columns are. 

You already mentioned a new crate where we can store database primitives. It is this crate but designed for `fuel-core` purposes. I will describe "`fuel-core` purposes" below.

## `fuel-core` purposes

The approach with ports and adaptors is good and will provide more flexibility for third parties who want to use our crates.

But at the same time, it means we move most of the logic into `fuel-core`. How to work with the storage is also a business logic of the service. Yes, it is not the responsibility of the service to know about the `RocksDB` or `InMemory` database, but it is the responsibility to update several storage tables, like `Coins` and `Messages`. 

We will face this problem while moving the `executor.rs` into the `fuel-executor` crate. [`ContractRef`](https://github.com/FuelLabs/fuel-core/blob/bad93d1e626aae61cfe9a8ede873f026508cc34b/fuel-block-executor/src/refs/contract.rs#L24) is an example of simplification development and aggregation of the business logic of the service. Moving this into `fuel-core` also transits the business logic of the service there.

My idea is to follow the ports/adapters approach but simultaneously provide a default implementation of the database port for `fuel-core`. Because our development is oriented toward `fuel-core` and support of modularity is an additional feature.

For third parties, it means that they have more ways of customizability:
1. They can reuse our `Database`(`RocksDB` or `InMemory`) and the default implementation of the database port then they have the same logic.
1. They can use their custom "database" but our default implementation of the port.
1. They can implement database port with their own logic.

Without it, they can only `1.` and `3.`.

At the same time, for us, we store the database business logic of the service inside of the service.

It is why it is database primitives for `fuel-core` purposes=)

## Other thoughts after the discussion above

Initially, I wanted to call this crate `fuel-storage`, but we already have the same crate. So I called it the `fuel-database`. But after some thought, it is the wrong naming. The proper name is `fuel-core-storage`.

The `fuel-storage` is a library for database primitives used by `fuel-core` and `fuel-vm`. Because `fuel-vm` is independent(doesn't know about `fuel-core`), it defines its database port. But also defines its `Mappable` types and provides a default implementation for this port.

The same with `fuel-core`. We use primitives from `fuel-storage` and `fuel-vm-storage` to define the required `Mappable` types for the default storage implementation around an unknown `Database`. So `fuel-core-storage` seems a good name, and we can use `fuel-database` for the crate with `Database` implementation(instead of storing it inside of the `fuel-core`).

The `fuel-core-storage` provides storage traits, mappable types, and other traits/types like `Transactional`/`Transaction<Database>`. It will be used by all services.

The `fuel-database` crate will provide an implementation of the `RocksDB`, `InMemory`, `Transactional<InMemory, RocksDB>` databases. It is used only by `fuel-core` in the `[dependencies]` section, and used by all service in the `[dev-dependencies]`

## Answers to questions

> Create an alternative to DatabaseTransaction that is entirely trait based. The new trait should have similar AsRef/Deref functionality as before allowing for casts to the db port of each respective service using it.

> I’ve read through this a few times and it’s not really landing. Can you provide a small example of what a trait based transaction could like like?

We discussed that yesterday on the call=) It is what I plan to implement [here](https://github.com/FuelLabs/fuel-core/compare/master...bugfix/race-condition-seal-block?expand=1) for the `poa-coordinator` service.
 
The idea is: 
Because the `DatabaseTransaction` structure is only known on the `fuel-core` level, we can't use it in the `fuel-block-producer`. But we can use a generic `DbTransaction: Transactional + AsRef<BlockProducerDB> + AsMut<BlockProducerDB>`, that implements transactional functionality(like `commit`) and also return a `Database` instance that implements the database port, defined by the by `fuel-block-producer` crate. 

With generics, it doesn't look lovely like:
```rust
pub trait Transactional {
    fn commit(self) -> anyhow::Result<()>;
}

// It is a database port defined by the `poa-coordiantor`.
pub trait BlockDb: Send + Sync {
    fn block_height(&self) -> Result<BlockHeight>;

    fn seal_block(
        &mut self,
        block_id: BlockId,
        consensus: FuelBlockConsensus,
    ) -> Result<()>;
}

#[async_trait::async_trait]
pub trait BlockProducer<DbTransaction, Database>: Send + Sync
where
    DbTransaction: Transactional + AsRef<Database> + AsMut<Database>,
    Database: BlockDb,
{
    // TODO: Right now production and execution of the block is one step, but in the future,
    //  `produce_block` should only produce a block without affecting the blockchain state.
    async fn produce_and_execute_block(
        &self,
        height: BlockHeight,
        max_gas: Word,
    ) -> anyhow::Result<UncommittedResult<DbTransaction>>;

    async fn dry_run(
        &self,
        transaction: Transaction,
        height: Option<BlockHeight>,
        utxo_validation: Option<bool>,
    ) -> anyhow::Result<Vec<Receipt>>;
}
```

You can see that `BlockProducer<DbTransaction, Database>` will cause the propagation of these two generics inside of `PoACoordinator`.

With `DatabaseTransaction<Database>` inside of `fuel-core-storage`, it would be `BlockProducer<Database>` it looks more friendly. Also, we plan to replace `dyn BlockDB` with `Database` in the future, so it is expected to be generic around `Database`=)

For now, instead of ugly `BlockProducer<DbTransaction, Database>`, we will use dynamic resolution:
```rust
pub trait Transactional {
    fn commit(self) -> anyhow::Result<()>;
}

// It is a database port defined by the `poa-coordiantor`.
pub trait BlockDb: Send + Sync {
    fn block_height(&self) -> Result<BlockHeight>;

    fn seal_block(
        &mut self,
        block_id: BlockId,
        consensus: FuelBlockConsensus,
    ) -> Result<()>;
}

pub trait DatabaseTransaction: Transactional + AsRef<&dyn BlockDb> + AsMut<&mut dyn BlockDb> {}

#[async_trait::async_trait]
pub trait BlockProducer: Send + Sync {
    // TODO: Right now production and execution of the block is one step, but in the future,
    //  `produce_block` should only produce a block without affecting the blockchain state.
    async fn produce_and_execute_block(
        &self,
        height: BlockHeight,
        max_gas: Word,
    ) -> anyhow::Result<UncommittedResult<Arc<dyn DatabaseTransaction>>>;

    async fn dry_run(
        &self,
        transaction: Transaction,
        height: Option<BlockHeight>,
        utxo_validation: Option<bool>,
    ) -> anyhow::Result<Vec<Receipt>>;
}
```

But this approach creates more `Arc` and `dyn` in our code, but doesn't require any ugly/not expected changes=)

Reply:
> The fuel-core-storage provides storage traits, mappable types, and other traits/types like Transactional/Transaction<Database>. It will be used by all services.
> 
> The fuel-database crate will provide an implementation of the RocksDB, InMemory, Transactional<InMemory, RocksDB> databases. It is used only by fuel-core in the [dependencies] section, and used by all service in the [dev-dependencies]
> 

Splitting into a fuel-core-storage vs fuel-database, where all the heavy deps and implementation logic of the database is separate, mostly resolves my concerns. As fuel-core-storage would just be high-level traits.

> You can see that BlockProducer<DbTransaction, Database> will cause the propagation of these two generics inside of PoACoordinator.

We definitely want to avoid propagating service-specific db traits between services, as it pollutes the domain with unrelated constraints.